File: AOC_PY/days/day13.py
import os
import numpy as np

INPUT_PATH = os.path.join(os.path.dirname(__file__), '../../Inputs', 'day13.txt')


class Day13:

    def __init__(self):
        with open(INPUT_PATH, 'r') as infile:
            lines = infile.read().splitlines()
        self.depart_time = int(lines[0])
        self.all_buses = lines[1].split(',')
        self.buses = [int(_) for _ in self.all_buses if _ != 'x']
        self.indices = [_i for _i, _a in enumerate(self.all_buses) if _a != 'x']

    def solve1(self):
        mods = [_ - self.depart_time % _ for _ in self.buses]
        pos = np.argmin(mods)
        return mods[pos] * self.buses[pos]

    def solve2(self):
        mods = [(x - i) % x for x, i in zip(self.buses, self.indices)]
        cur, cm = 1, 1
        for val, m in zip(self.buses, mods):
            while cur % val != m:
                cur += cm
            cm = np.lcm(cm, val)
        return cur


def main():
    x = Day13()
    print(x.solve1())
    print(x.solve2())


if __name__ == '__main__':
    main()

File: AOC_CS/Days/Day13.cs
using System.IO;
using System.Linq;
using AOC_CS.Utils;
using AOC_CS.Utils.Extensions;

namespace AOC_CS.Days
{
    class Day13
    {
        const string INPUT_FILE = "Inputs/day13.txt";

        private int departTime;
        private string[] allBuses;
        private int[] buses;
        private int[] indices;

        public Day13()
        {
            departTime = 0;
            allBuses = null;
            buses = null;
        }

        private void ReadInput()
        {
            if (indices != null) return;
            var lines = File.ReadAllLines(INPUT_FILE);
            departTime = int.Parse(lines[0]);
            allBuses = lines[1].Split(',');
            buses = allBuses.Where(x => x != "x").Select(int.Parse).ToArray();
            indices = allBuses.Select((x, i) => x == "x" ? -1 : i).Where(x => x >= 0).ToArray();
        }

        public int Solve1()
        {
            ReadInput();
            var mods = buses.Select(x => x - departTime % x).ToArray();

            var pos = mods.ArgMin();
            return mods[pos] * buses[pos];
        }

        public long Solve2()
        {
            ReadInput();
            var mods = buses.Select((x, i) =>
            {
                var tmp = (x - indices[i]) % x;
                return tmp < 0 ? tmp + x : tmp;
            }).ToArray();

            long l = 1, cur = 1;
            for (int i = 0; i < buses.Length; ++i)
            {
                while (cur % buses[i] != mods[i])
                    cur += l;
                l = Numbers.Lcm(l, buses[i]);
            }
            return cur;
        }
    }
}
